npm/npm_import.bzl

```diff
@@ -204,35 +206,55 @@ def npm_translate_lock(
     Args:
         name: The repository rule name
 
-        pnpm_lock: The pnpm-lock.yaml file.
+        pnpm_lock: The `pnpm-lock.yaml` file.
+
+            Must be set if `npm_package_lock` and `yarn_lock` are not set.
+
+        pnpm_workspace: The pnpm-workspace.yaml file, if any.
 
-            Exactly one of [pnpm_lock, npm_package_lock, yarn_lock] should be set.
+            Should be set if there are any non-root packages in the pnpm-workspace.
 
-        package_json: The package.json file. From this file and the corresponding `package-lock.json`/`yarn.lock` file
-            (specified with the `npm_package_lock`/`yarn_lock` attributes),
-            a `pnpm-lock.yaml` file will be generated using `pnpm import`.
+        npm_package_lock: The `package-lock.json` file written by `npm install`.
 
-            Note that *any* changes to the package.json file will invalidate the npm_translate_lock
-            repository rule, causing it to re-run on the next invocation of Bazel.
+            Only one of [npm_package_lock, yarn_lock] may be set.
 
-            Mandatory when using npm_package_lock or yarn_lock, otherwise must be unset.
+        yarn_lock: The `yarn.lock` file written by `yarn install`.
 
-        npm_package_lock: The package-lock.json file written by `npm install`.
+            Only one of [npm_package_lock, yarn_lock] may be set.
 
-            When set, the `package_json` attribute must be set as well.
-            Exactly one of [pnpm_lock, npm_package_lock, yarn_lock] should be set.
+        update_pnpm_lock: When True, the pnpm lock file wil be updated automatically when it is out-of-date.
 
-        yarn_lock: The yarn.lock file written by `yarn install`.
+            Defaults to False unless `npm_package_lock` or `yarn_lock` are set, in which case it defaults
+            to True.
 
-            When set, the `package_json` attribute must be set as well.
-            Exactly one of [pnpm_lock, npm_package_lock, yarn_lock] should be set.
+            A `pnpm-lock.rules_js` file will be created next to `pnpm-lock.yaml` and used to determine when
```

Comment:
I wonder if we can add metadata to something such as the lockfile itself? If we control the lockfile can we put comments at the top?

Reply:
Suppose we could but it doesn't seem ideal; for one running `pnpm i --lockfile-only` will remove anything we add.

I'm thinking of putting the "hash" file into an `.aspect/external_repository_hashes/` folder so it doesn't pollute the root of the repository